server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterInfoSimulatorTests.java

```diff
@@ -133,6 +153,191 @@ public void testRelocateShardWithMultipleDataPath1() {
+            SnapshotShardSizeInfo.EMPTY,
+            0L
+        );
+        var routingNodes = allocation.routingNodes();
+
+        assertThat(
+            "Should keep index-1 on node-0",
+            decider.canRemain(clusterState.metadata().index("index-1"), shard1, routingNodes.node("node-0"), allocation).type(),
+            equalTo(Decision.Type.YES)
+        );
+        assertThat(
+            "Should keep index-2 on node-0",
+            decider.canRemain(clusterState.metadata().index("index-2"), shard2, routingNodes.node("node-0"), allocation).type(),
+            equalTo(Decision.Type.YES)
+        );
+        assertThat(
+            "Should not allocate index-3 on node-0 (not enough space)",
+            decider.canAllocate(shard3, routingNodes.node("node-0"), allocation).type(),
+            equalTo(Decision.Type.NO)
+        );
+    }
+
+    public void testDiskUsageSimulationWithMultipleDataPathAndDiskThresholdDecider() {
+
+        var discoveryNodesBuilder = DiscoveryNodes.builder()
+            .add(createDiscoveryNode("node-0", DiscoveryNodeRole.roles()))
+            .add(createDiscoveryNode("node-1", DiscoveryNodeRole.roles()))
+            .add(createDiscoveryNode("node-2", DiscoveryNodeRole.roles()));
+
+        var metadataBuilder = Metadata.builder();
+        var routingTableBuilder = RoutingTable.builder();
+
+        var shard1 = newShardRouting("index-1", 0, "node-0", null, true, STARTED);
+        addIndex(metadataBuilder, routingTableBuilder, shard1);
+
+        var shard2 = newShardRouting("index-2", 0, "node-0", "node-1", true, INITIALIZING);
+        addIndex(metadataBuilder, routingTableBuilder, shard2);
+
+        var shard3 = newShardRouting("index-3", 0, "node-1", null, true, STARTED);
+        addIndex(metadataBuilder, routingTableBuilder, shard3);
+
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(discoveryNodesBuilder)
+            .metadata(metadataBuilder)
+            .routingTable(routingTableBuilder)
+            .build();
+
+        var simulator = new ClusterInfoSimulator(
+            new ClusterInfoTestBuilder() //
+                .withNode("node-0", new DiskUsageBuilder("/data-1", 1000, 100), new DiskUsageBuilder("/data-2", 1000, 500))
+                .withNode("node-1", new DiskUsageBuilder("/data-1", 1000, 100), new DiskUsageBuilder("/data-2", 1000, 300))
+                .withShard(shard1, 500)
+                .withShard(shard2, 400)
+                .withShard(shard3, 300)
+                .build()
+        );
+
+        simulator.simulate(shard2);
+
+        assertThat(
+            simulator.getClusterInfo(),
+            equalTo(
+                new ClusterInfoTestBuilder() //
+                    .withNode("node-0", new DiskUsageBuilder("/data-1", 1000, 0), new DiskUsageBuilder("/data-2", 1000, 500))
+                    .withNode("node-1", new DiskUsageBuilder("/data-1", 1000, 100), new DiskUsageBuilder("/data-2", 1000, 700))
```

Comment:
shard2 with size 400 is moved from node-1 to node-0. This result does not seems to be correct. I am working on fixing this.